[torchbind] fix bug of mutating FakeScriptObjects twice in aot_export

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128844

This PR does two things:
1. it duplicates the fake script object because aot_export trace the program twice. The result of tracing in the first time would cause the tracing result of second time be wrong.
2. Also add a new test for methods that return constant outputs. Before the PR, there's is no meta["val"] for these nodes because fx won't track these constants. We still need to preserve these constant return operators in the graph because torchbind objects are stateful and deleting it would remove the implicit state mutation inside of the object.

ydwu4 committed:
[torchbind] fix bug of mutating FakeScriptObjects twice in aot_export

[ghstack-poisoned]

ydwu4 committed:
Update on "[torchbind] fix bug of mutating FakeScriptObjects twice in aot_export"


This PR does two things:
1. it duplicates the fake script object because aot_export trace the program twice. The result of tracing in the first time would cause the tracing result of second time be wrong.
2. Also add a new test for methods that return constant outputs. Before the PR, there's is no meta["val"] for these nodes because fx won't track these constants. However, we need to preserve these constant return operators in the graph because torchbind objects are stateful and deleting it would remove the implicit state mutation inside of the object.



[ghstack-poisoned]

angelayi reviewed (approved):
I'm not sure if it's good to store the real script object inside of the fake script object.. because it's like, what's the point of having a fake one then 😂 
But I can't think of a better way to duplicate a FakeScriptObject so I will stamp

ydwu4 committed:
Update on "[torchbind] fix bug of mutating FakeScriptObjects twice in aot_export"


This PR does two things:
1. it duplicates the fake script object because aot_export trace the program twice. The result of tracing in the first time would cause the tracing result of second time be wrong.
2. Also add a new test for methods that return constant outputs. Before the PR, there's is no meta["val"] for these nodes because fx won't track these constants. We still need to preserve these constant return operators in the graph because torchbind objects are stateful and deleting it would remove the implicit state mutation inside of the object.



[ghstack-poisoned]

ydwu4 committed:
Update on "[torchbind] fix bug of mutating FakeScriptObjects twice in aot_export"


This PR does two things:
1. it duplicates the fake script object because aot_export trace the program twice. The result of tracing in the first time would cause the tracing result of second time be wrong.
2. Also add a new test for methods that return constant outputs. Before the PR, there's is no meta["val"] for these nodes because fx won't track these constants. We still need to preserve these constant return operators in the graph because torchbind objects are stateful and deleting it would remove the implicit state mutation inside of the object.



[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
